Sources/TuistCore/Graph/Graph.swift

```diff
@@ -240,9 +240,12 @@ public class Graph: Encodable, Equatable {
 
         // Precompiled libraries and frameworks
 
-        let precompiledLibrariesAndFrameworks = targetNode.precompiledDependencies
-            .lazy
-            .map(GraphDependencyReference.init)
+        let precompiledLibrariesAndFrameworks = findAll(
```

Comment:
I think I have a better idea of why this change is necessary. If I'm not wrong, the issue is that transitive pre-compiled dependencies are not linked, right? I think we might need to adjust your changes a bit further. We might have the following scenarios:

1. App ---(depends on)---> Precompiled static binary (A) ---> Precompiled static binary (B)
2. App ---(depends on)---> Precompiled dynamic binary (A) ----> Precompiled dynamic binary (B)
3. App ---(depends on)---> Precompiled static binary (A) ----> Precompiled dynamic binary (B)

I think what we should get at the app level is:
1. App links both binaries
2. App links A, and copies A and B into the product.
3. App links A, and copies B into the product.

If I'm not wrong, with your changes in the 3rd case App would be linking B, which is not what we want because that'll create an implicit dependency between the app and B and users will be able to import B.

Notice that `findAll` has 2 arguments, `test`, and `skip` that you can use to select the right dependencies here:

```swift
findAll(targetNode: targetNode, test: and(isStatic, isPrecompiled), skip: canLinkStaticProducts)
```

@kwridan am I missing something? 

I think we didn't have this scenario before caching because we didn't have the notion of dependencies between pre-compiled nodes. However, that's a thing now and we should account for it.

Reply:
> I think we didn't have this scenario before caching because we didn't have the notion of dependencies between pre-compiled nodes. However, that's a thing now and we should account for it.

Indeed, precompiled frameworks depending on one another is an internal detail used for `tuist cache` and not something exposed during non-cached workflows.

To cater for the scenarios above perhaps something less eager is needed. I believe @mollyIV first attempt at this was more correct:

https://github.com/tuist/tuist/pull/2051/commits/c317a709a30dfdc7cd55b459c12bb47b5c048d36

The latest change here will transitively include ALL transitive precompiled nodes:

- Target A > Precompiled A
- Target A > Target B > Precompiled B

This change would result in TargetA ending up with `Precompiled A` and `Precompiled B` which is not the correct behaviour in all circumstances (we have that covered [in the next set of lines below](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Graph/Graph.swift#L249))


We can see the impact/change in behaviour in the fixture `ios_app_with_transitive_framework`: 
- generate the projects (without caching) using latest tuist and this PR :/ 

This PR:

![Screenshot 2020-12-07 at 10 12 22](https://user-images.githubusercontent.com/11914919/101338323-b93bd500-3874-11eb-97fa-ca48fa9d201b.png)

![Screenshot 2020-12-07 at 10 09 34](https://user-images.githubusercontent.com/11914919/101338247-9f01f700-3874-11eb-8e2e-2929f10bf933.png)

![Screenshot 2020-12-07 at 10 09 27](https://user-images.githubusercontent.com/11914919/101338242-9e696080-3874-11eb-83c2-9f962b2500cf.png)

Latest tuist release _(correct behaviour)_:

![Screenshot 2020-12-07 at 10 10 02](https://user-images.githubusercontent.com/11914919/101338396-d2dd1c80-3874-11eb-9ff6-2bbb004a1c71.png)

![Screenshot 2020-12-07 at 10 13 41](https://user-images.githubusercontent.com/11914919/101338483-f011eb00-3874-11eb-8fd7-baa4928be3a0.png)

![Screenshot 2020-12-07 at 10 13 51](https://user-images.githubusercontent.com/11914919/101338500-f607cc00-3874-11eb-8b13-b71876f1da39.png)

Reply:
Follow up from slack discussions:

- The recursive precompiled dependencies may need to be augmented with some additional logic as suggested by @pepibumur 

We'll need to ensure we traverse the `PrecompiledNode`s path as opposed the `TargetNode`s - its a little tricky as the `findAll` was geared towards `TargetNode` traversal.

![precompiled-nodes](https://user-images.githubusercontent.com/11914919/101348133-0f177980-3883-11eb-914f-769bc70ded11.png)


Something like this may do the trick:


```swift
let isPrecompiled: (GraphNode) -> Bool = { $0 is PrecompiledNode }
let isDynamicPrecompiledNode: (PrecompiledNode) -> Bool = { $0.isDynamicAndLinkable() }
let isTargetNode: (GraphNode) -> Bool = { $0 is TargetNode }
let precompiledLibrariesAndFrameworks = findAll(targetNode: targetNode, test: isPrecompiled, skip: or(isDynamicPrecompiledNode, isTargetNode))
        .lazy
        .map(GraphDependencyReference.init)
```

Yes in some cases we do need to care about target transitive precompiled nodes, but that [logic is already in place](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Graph/Graph.swift#L249) :)

We can try to formalise those into unit tests within the `Graph` and `ValueGraphTraverser`

Scenarios:

**1. App ---(depends on)---> Precompiled static binary (A) ---> Precompiled static binary (B)**

Linkable: 
- Precompiled static binary (A)
- Precompiled static binary (B)

Embeddable: _None_

**2. App ---(depends on)---> Precompiled dynamic binary (A) ----> Precompiled dynamic binary (B)**

Linkable: 
- Precompiled dynamic binary (A)

Embeddable:
- Precompiled dynamic binary (A)
- Precompiled dynamic binary (B)

**3. App ---(depends on)---> Precompiled static binary (A) ----> Precompiled dynamic binary (B)**

Linkable: 
- Precompiled static binary (A)
- Precompiled dynamic binary (B)

Embeddable:
- Precompiled dynamic binary (B)

**4. App ---(depends on)---> Precompiled dynamic binary (A) ----> Precompiled static binary (B)**

Linkable: 
- Precompiled dynamic binary (A)

Embeddable:
- Precompiled dynamic binary (A)

Notes:
- In this case Precompiled static binary (B) has already been linked into Precompiled dynamic binary (A) - as such no further action is needed

**5. App ---(depends on)---> Precompiled static binary (A)**

Linkable: 
- Precompiled static binary (A)

Embeddable: None

**6. App ---(depends on)---> Precompiled dynamic binary (A)**

Linkable: 
- Precompiled dynamic binary (A)

Embeddable:
- Precompiled dynamic binary (A)